src/java.base/share/classes/com/sun/crypto/provider/HmacCore.java

```diff
@@ -58,7 +65,7 @@ abstract class HmacCore extends MacSpi implements Cloneable {
      */
     HmacCore(String digestAlgo, int bl) throws NoSuchAlgorithmException {
         MessageDigest md = MessageDigest.getInstance(digestAlgo);
-        if (!(md instanceof Cloneable)) {
+        if (!(md instanceof Cloneable) && !RestrictedSecurity.isFIPSEnabled()) {
```

Comment:
While we enabled {Mac, HmacPBESHA256, *} in provider SunJCE in java.security file, we hit an error which states that `no such algorithm: SHA-256 for provider SUN`. 

This issue, `no such algorithm: SHA-256 for provider SUN`, happens at `HmacCore.java`, when using the "Mac", "HmacPBESHA256" in the provider SunJCE, it will go into the implementation class `com.sun.crypto.provider.HmacPKCS12PBECore`.

```
ps("Mac", "HmacPBESHA256",
                "com.sun.crypto.provider.HmacPKCS12PBECore$HmacPKCS12PBE_SHA256",
                null, attrs);
```
And in this `HmacPKCS12PBECore` construction method, it will call its super class, which is the class `HmacCore`, its construction method. And due to the `Mac` is `HmacPBESHA256`, then it will call `HmacCore` construction method with algorithm `SHA-256`.

And in the `HmacCore` construction method, below. If the `MessageDigest` is not instance of `cloneable`, then the MessageDigest will be initialized from provider `SUN`.
```
HmacCore(String digestAlgo, int bl) throws NoSuchAlgorithmException {
        MessageDigest md = MessageDigest.getInstance(digestAlgo);
        if (!(md instanceof Cloneable)) {
            Provider sun = Security.getProvider("SUN");
            if (sun != null) {
                md = MessageDigest.getInstance(digestAlgo, sun);
            ……
```

In our case, the MessageDigest will be initialized using algorithm `SHA-256` from provider `SunPKCS11-NSS-FIPS`, by its MessageDigest implementation class `P11Digest`, and it is an instance of `Cloneable`.
```
final class P11Digest extends MessageDigestSpi implements Cloneable,
    MessageDigestSpi2 {
……
```

But, in the HmacCore construction method, the `if (!(md instanceof Cloneable))` still return true, which mean, `MessageDigest` from `SunPKCS11-NSS-FIPS` is not detected as `cloneable`. But actually, from the source code, above, it is cloneable.

But again, the same codes return cloneable in JDK17 and above. we did some tests, test codes below.
```
public static void main(String[] args) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        System.out.println("Tao debug - provider: " + md.getProvider());
        System.out.println("Tao debug - cloneable: " + (md instanceof Cloneable));
    }
```
And run this test code using JDK11, 17 and 20, the results below.
for 11:
```
java -Dsemeru.restrictedsecurity=1 com.test.fips.testtest
Tao debug - provider: SunPKCS11-NSS-FIPS version 11
Tao debug - cloneable: false
```
for 17:
```
java -Dsemeru.restrictedsecurity=1 com.test.fips.testtest
Tao debug - provider: SunPKCS11-NSS-FIPS version 17
Tao debug - cloneable: true
```
for 20:
```
java -Dsemeru.restrictedsecurity=1 com.test.fips.testtest
Tao debug - provider: SunPKCS11-NSS-FIPS version 20
Tao debug - cloneable: true
```
The same code, initialize the `MessageDigest` from provider `SunPKCS11-NSS-FIPS`, by its MessageDigest implimetation class `P11Digest`. In JDK11 return not cloneable, but in JDK17 and above return cloneable.

Not sure it is a JDK issue or not, but looks like it is fixed by JDK17.

So, for our case, we suggest, for JDK11 and maybe also for JDK8, we change the `if (!(md instanceof Cloneable)) `to `if (!(md instanceof Cloneable) && !RestrictedSecurityConfigurator.isFIPSSupportPKCS12()) `. So, in this, the MessageDigest will be initialized by using provider SunPKCS11-NSS-FIPS, not go into the cloneable check and then initialized by using the hard code provider SUN.

Reply:
This change belongs in a separate pull request (assuming it's still needed given this removes the SunJCE provider for restricted mode).

Reply:
I don't understand the need to change this at all. As you point out, `P11Digest implements Cloneable` so the code path would not be taken anyway. Even for non-cloneable objects, the code anticipates that the "SUN" provider might not be available and compensates.

Reply:
Let me explain this, yes, the P11Digest class does implement Cloneable. But in jdk11, it seems that the MessageDigest object initialized from provider SunPKCS11-NSS-FIPS is not cloneable, by its MessageDigest implimetation class P11Digest. The following code is the sample test.
```
public static void main(String[] args) throws Exception { 
    MessageDigest md = MessageDigest.getInstance("SHA-256");
    System.out.println("Tao debug - provider: " + md.getProvider());
    System.out.println("Tao debug - cloneable: " + (md instanceof Cloneable)); 
}
```
Running this samples test in jdk11, jdk17, and jdk20 return different results.
for 11:
```
java -Dsemeru.restrictedsecurity=1 com.test.fips.testtest
Tao debug - provider: SunPKCS11-NSS-FIPS version 11
Tao debug - cloneable: false
```
for 17:
```
java -Dsemeru.restrictedsecurity=1 com.test.fips.testtest
Tao debug - provider: SunPKCS11-NSS-FIPS version 17
Tao debug - cloneable: true
```
for 20:
```
java -Dsemeru.restrictedsecurity=1 com.test.fips.testtest
Tao debug - provider: SunPKCS11-NSS-FIPS version 20
Tao debug - cloneable: true
```
As you can see, in jdk11, the MessageObject is not cloneable. And we do not want to use the SHA-256 from the SUN provider in fips mode, we want it from the SunPKCS11-NSS-FIPS provider. That's the reason why we make this change.

Reply:
The difference you note between Java 11 and Java 17 seems to have originated with
* [8246077: Cloneable test in HmacCore seems questionable](https://github.com/ibmruntimes/openj9-openjdk-jdk17/commit/b88cc7219e1bd53bda2adefa8cfffb3bfbc0f8f9)

I'm still not convinced the change you propose to this file is appropriate, but it definitely does not belong hidden in this pull request. If you believe something should change with respect to message digests in restricted security mode, please create an issue where we can discuss how to address it.

But while we're discussing this, in `java.security`, `RestrictedSecurity1.jce.provider.2 = SUN ...` should not allow that provider to offer SHA-256 (used in your example): As I read it, it will not, but if I'm wrong, that's where I think the fix should be.

Reply:
Let me open an issue for this.

Reply:
Please share here a link to that issue.

Reply:
https://github.com/ibmruntimes/openj9-openjdk-jdk11/issues/665

Reply:
1. We dont enable `MessageDigest` in SUN provider, this means such `SHA-256` messagedigest instance should comes from `SunPKCS11-NSS-FIPS` provider in FIPS mode. 

2. The problem is that since MessageDigest if not cloneable, and this code
`if (!(md instanceof Cloneable)) {`
 will return true, and then, the code will go forward to get the MessageDigest instance from the SUN provider.
```
            Provider sun = Security.getProvider("SUN");
            if (sun != null) {
                md = MessageDigest.getInstance(digestAlgo, sun);
```
 But we dont enable `MessageDigest` in SUN provider. And this will cause the `no such algorithm: SHA-256 for provider SUN` error.

Reply:
This still feels like a separate issue. Please revert the change to `HmacCore.java` here and we'll figure out the right course of action based on a discussion that should occur in #665.

Reply:
Reverted